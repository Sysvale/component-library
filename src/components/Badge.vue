<template>
	<div
		class="badge-container"
		:style="colorCodeMode ? styleVariables : ''"
		:class="!colorCodeMode ? predefinedStyle : ''"
	>
		<div
			class="d-flex justify-space-between align-items-center padding-left-12"
		>
			<small class="regular">
				{{ content }}
			</small>

			<div
				class="dismissible-icon-container ml-1 dismissible-icon-container d-flex"
				:class="isDismissible ? '' : 'padding-right-12'"
				@click="$emit('close', true)"
			>
				<x-icon
					v-if="isDismissible"
					size="0.9x"
					class="m-1"
				/>
			</div>
		</div>
	</div>
</template>

<script>
import { XIcon } from 'vue-feather-icons'
export default {
	components: {
		XIcon,
	},
	
	props: {
		content: {
			type: String,
			default: 'Negativo',
			description: 'The text that will be displayed inside the badge.',
			required: true,
		},
		color: {
			type: String,
			default: '#28A745',
			description: `If the color code mode is set,
				you can pass the hex value of the text
				color of the badge.`,
		},
		bgColor: {
			type: String,
			default: '#28a74526',
			description: `If the color code mode is set,
				you can pass the hex value of the background
				color of the badge.`,
		},
		colorCodeMode: {
			type: Boolean,
			default: false,
			description: `The property that specifies 
				if the badge will work in the color code
				mode or not.`,
		},
		variant: {
			type: String,
			default: 'yellow',
			description: `Variant mode gives 6 predefined badges 
				to work with. Seting this property to 'true',
				you can use 'yellow', 'green', ... 'gray'
				to use the predefined badge styles.`,
		},
		isDismissible: {
			type: Boolean,
			default: false,
			description: `Specifies if the badge will display a 
				'x' icon that emits the 'close' event.`,
		},
	},

	computed: {
		predefinedStyle() {
			let computed_style = '';
			switch (this.variant) {
				case 'yellow':
					computed_style = 'yellow';
					break;
				case 'green':
					computed_style = 'green';
					break;
				case 'red':
					computed_style = 'red';
					break;
				case 'blue':
					computed_style = 'blue';
					break;
				case 'purple':
					computed_style = 'purple';
					break;
				case 'gray':
					computed_style = 'gray';
					break;
				default:
					computed_style = 'yellow';
					break;
			}

			return computed_style;
		},

		styleVariables() {
			return {
				'--bg-color': this.bgColor,
				'--color': this.color,
			};
		},
	},
};
</script>
<style>
.badge-container {
	border-radius: 8px !important;
	width: fit-content;
	background-color: var(--bg-color);
	color: var(--color);
	padding: 2px 4px;
}

.yellow {
	color: #5a4300;
	background-color: #ffbf004d;
}

.green {
	color: #005a15;
	background-color: #28a74526;
}

.red {
	color: #8c1520;
	background-color: #e2757f40;
}

.blue {
	color: #00469c;
	background-color: #60c4ff40;
}

.purple {
	color: #621e83;
	background-color: #e47aff40;
}

.gray {
	color: #2f2f2f;
	background-color: #acacac40;
}

.padding-right-12 {
	padding-right: 12px;
}

.padding-left-12 {
	padding-left: 12px;
}

.dismissible-icon-container {
	cursor: pointer;
	border-radius: 100%;
}

.dismissible-icon-container:hover {
	background-color: hsla(0, 0%, 10%, 0.1);
	transition: 0.5s ease;
}
</style>
