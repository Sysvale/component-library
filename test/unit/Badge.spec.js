import { mount, createLocalVue } from '@vue/test-utils';
import BootstrapVue from 'bootstrap-vue';
import Badge from '../../src/components/Badge.vue';

const localVue = createLocalVue();
localVue.use(BootstrapVue);

test('Component is mounted properly', () => {
	const wrapper = mount(Badge, {
		localVue,
		propsData: {
			content: 'Test',
		},
	});
	expect(wrapper).toMatchSnapshot();
});

describe("Computed property 'predefinedStyle' test", () => {
	test('if the computed property predefinedStyle works properly with the default value for the prop variant', () => {
		const wrapper = mount(Badge, {
			localVue,
			propsData: {
				content: 'Test',
			},
		});
		expect(wrapper.vm.predefinedStyle).toBe('yellow');
	});
});

describe("Prop 'variant' tests", () => {
	test("if the computed property changes when the prop variant is setted to 'yellow'", () => {
		const wrapper = mount(Badge, {
			localVue,
			propsData: {
				variant: 'yellow',
				content: 'Test',
			},
		});

		expect(wrapper.vm.predefinedStyle).toBe('yellow');
	});

	test("if the computed property changes when the prop variant is setted to 'green'", () => {
		const wrapper = mount(Badge, {
			localVue,
			propsData: {
				variant: 'green',
				content: 'Test',
			},
		});

		expect(wrapper.vm.predefinedStyle).toBe('green');
	});

	test("if the computed property changes when the prop variant is setted to 'red'", () => {
		const wrapper = mount(Badge, {
			localVue,
			propsData: {
				variant: 'red',
				content: 'Test',
			},
		});

		expect(wrapper.vm.predefinedStyle).toBe('red');
	});

	test("if the computed property changes when the prop variant is setted to 'blue'", () => {
		const wrapper = mount(Badge, {
			localVue,
			propsData: {
				variant: 'blue',
				content: 'Test',
			},
		});

		expect(wrapper.vm.predefinedStyle).toBe('blue');
	});

	test("if the computed property changes when the prop variant is setted to 'purple'", () => {
		const wrapper = mount(Badge, {
			localVue,
			propsData: {
				variant: 'purple',
				content: 'Test',
			},
		});

		expect(wrapper.vm.predefinedStyle).toBe('purple');
	});

	test("if the computed property changes when the prop variant is setted to 'gray'", () => {
		const wrapper = mount(Badge, {
			localVue,
			propsData: {
				variant: 'gray',
				content: 'Test',
			},
		});

		expect(wrapper.vm.predefinedStyle).toBe('gray');
	});
});

describe("Prop 'bgColor' and 'color' tests", () => {
	test("if the computed property 'styleVariables' changes when the prop 'colorCodeMode' is setted to 'true'", () => {
		const wrapper = mount(Badge, {
			localVue,
			propsData: {
				colorCodeMode: true,
				color: 'white',
				bgColor: 'blue',
				content: 'Test',
			},
		});

		expect(wrapper.vm.styleVariables).toStrictEqual({"--bg-color": "blue", "--color": "white"});
	});
});

test('if the event is emited correctly when the dismissible icon is clicked', () => {
	const wrapper = mount(Badge, {
		localVue,
		propsData: {
			isDismissible: true,
			content: 'Test',
		},
	});

	wrapper.find('.dismissible-icon-container').trigger('click');

	expect(wrapper.emitted().close).toBeTruthy();
	expect(wrapper.emitted().close).toEqual([ [ true ] ]);
});