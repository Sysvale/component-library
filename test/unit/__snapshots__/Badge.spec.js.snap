// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component is mounted properly 1`] = `
<div class="badge-container yellow">
  <div class="d-flex justify-space-between align-items-center padding-left-12"><small class="regular">
      Test
    </small>
    <div class="dismissible-icon-container ml-1 d-flex padding-right-12">
      <!---->
    </div>
  </div>
</div>
`;
